lib/finalize-manifest.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+var checksumStream = require('checksum-stream')
+var dezalgo = require('dezalgo')
+var finished = require('mississippi').finished
+var gunzip = require('./util/gunzip-maybe')
+var minimatch = require('minimatch')
+var normalize = require('normalize-package-data')
+var optCheck = require('./util/opt-check')
+var path = require('path')
+var pipe = require('mississippi').pipe
+var pipeline = require('mississippi').pipeline
+var tar = require('tar-stream')
+var through = require('mississippi').through
+
+module.exports = finalizeManifest
+function finalizeManifest (pkg, spec, opts, cb) {
+  tarballedProps(pkg, spec, opts, function (err, props) {
+    if (err) { return cb(err) }
+    // normalize should not add any fields, and once
+    // makeManifest completes, it should never be modified.
+    cb(null, new Manifest(pkg, props))
+  })
+}
+
+module.exports.Manifest = Manifest
+function Manifest (pkg, fromTarball) {
+  fromTarball = fromTarball || {}
+  this.name = pkg.name
+  this.version = pkg.version
+  this.dependencies = pkg.dependencies || {}
+  this.optionalDependencies = pkg.optionalDependencies || {}
+  this.devDependencies = pkg.devDependencies || {}
+  var bundled = (
+    pkg.bundledDependencies ||
+    pkg.bundleDependencies ||
+    false
+  )
+  this.bundleDependencies = bundled
+  this.peerDependencies = pkg.peerDependencies || {}
+
+  // This one depends entirely on each handler.
+  this._resolved = pkg._resolved
+
+  // We sometimes need to grab these from the tarball.
+  this._shasum = pkg._shasum || fromTarball._shasum
+  this._shrinkwrap = pkg._shrinkwrap || fromTarball._shrinkwrap || null
+  this.bin = pkg.bin || fromTarball.bin || null
+
+  if (this.bin && Array.isArray(this.bin)) {
+    // This isn't even documented?
+    // Code yanked from read-package-json
+    var m = pkg.directories && pkg.directories.bin || '.'
+    this.bin = this.bin.reduce(function (acc, mf) {
+      if (mf && mf.charAt(0) !== '.') {
+        var f = path.basename(mf)
+        acc[f] = path.join(m, mf)
+      }
+      return acc
+    }, {})
+  }
+
+  this._id = null // filled in by normalize-package-data, but unnecessary
+
+  // TODO - freezing and inextensibility pending npm changes. See test suite.
+  // Object.preventExtensions(this)
+  normalize(this)
+
+  // I don't want this why did you give it to me. Go away. 🔥🔥🔥🔥
+  delete this.readme
+
+  // Object.freeze(this)
```

Comment:
This is a reminder to myself of something I really want to do but npm doesn't let me do yet -- it'll be incommented once I figure a few things out, and I didn't want it to be forgotten.